futures-util/src/future/poll_immediate.rs

```diff
@@ -0,0 +1,201 @@
+use super::assert_future;
+use crate::FutureExt;
+use core::pin::Pin;
+use futures_core::task::{Context, Poll};
+use futures_core::{FusedFuture, Future, Stream};
+
+/// Future for the [`poll_immediate`](poll_immediate()) function.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct PollImmediate<T>(Option<T>);
+
+impl<T, F> Future for PollImmediate<F>
+where
+    F: Future<Output = T>,
+{
+    type Output = Option<T>;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        // # Safety
+        // This is the only time that this future will ever be polled.
+        let inner =
+            unsafe { self.get_unchecked_mut().0.take().expect("PollOnce polled after completion") };
```

Comment:
this is unsound when `Future::poll` called after `fut.poll()` (in `Stream` impl) panicked.

Reply:
I've removed all unsafe code, so hopefully this is now sound.